### PR TITLE
Implement Java version of the stable marriage problem.

### PR DESCRIPTION
**Pull request by bsamseth**

### CODE REVIEW

File: contents/stable_marriage_problem/code/java/stable-marriage.java [hunk 1]
```diff
@@ -0,0 +1,146 @@
...
+import java.util.Collections;
+
+class StableMarriage {
+
+    /*
+     * Use the stable marriage algorithm to find stable pairs from the
+     * lists of men and women.
+     */
+    public static void findStableMarriages(List<Woman> women, List<Man> men) {
+        boolean keep_going;
```
Comment (Gustorn) [inline contents/stable_marriage_problem/code/java/stable-marriage.java]: Going by the Java style guide this should be `keepGoing`

File: contents/stable_marriage_problem/code/java/stable-marriage.java [hunk 2]
```diff
@@ -0,0 +1,146 @@
...
+                    man.propose();
+
+            // The women pick their favorite suitor.
+            for (Woman woman : women)
+                woman.chooseMate();
+
+            // Repeat the process if someone is left without a partner.
+            keep_going = false;
+            for (Person person : leastCommonGender) {
+                if (keep_going |= person.isLonely()) {
```
Comment (Gustorn) [inline contents/stable_marriage_problem/code/java/stable-marriage.java]: I really don't like this. First, it uses the less common variant of the OR operator (non short-circuiting). Second, it checks the result of an assignemnt which is probably even worse. You could do this instead:

```java
keepGoing = leastCommonGender.stream().anyMatch(Person::isLonely)
```

Or alternatively, get rid of `keepGoing` altogether:

```java
if (!leastCommonGender.anyMatch(Person::isLonely)) {
    break;
}
```

File: contents/stable_marriage_problem/code/java/stable-marriage.java [hunk 3]
```diff
@@ -0,0 +1,146 @@
...
+            }
+
+        } while (keep_going);
+
+        women.forEach(w -> System.out.println(w + " married to " + w.getMate()));
+    }
+
+    public static void main(String[] args) {
+        int nPairs = 5;
+        List<Woman> women = new LinkedList<>();
```
Comment (Gustorn) [inline contents/stable_marriage_problem/code/java/stable-marriage.java]: You should really use `ArrayList` in both of these cases. `LinkedList` is just a bad data structure in general

File: contents/stable_marriage_problem/code/java/stable-marriage.java [hunk 4]
```diff
@@ -0,0 +1,146 @@
...
+
+        findStableMarriages(women, men);
+    }
+
+}
+
+class Person {
+    private final String name;
+    protected Person mate;
+    protected List<Person> preferedMates;
```
Comment (Gustorn) [inline contents/stable_marriage_problem/code/java/stable-marriage.java]: prefered -> preferred

File: contents/stable_marriage_problem/code/java/stable-marriage.java [hunk 5]
```diff
@@ -0,0 +1,146 @@
...
+        }
+    }
+
+    public Person getMate() {
+        return mate;
+    }
+
+    public void receiveOptions(List<? extends Person> mates) {
+        // Preferences are subjective.
+        preferedMates = new LinkedList<>(mates);
```
Comment (Gustorn) [inline contents/stable_marriage_problem/code/java/stable-marriage.java]: Similarly, use an ArrayList here

File: contents/stable_marriage_problem/code/java/stable-marriage.java [hunk 6]
```diff
@@ -0,0 +1,146 @@
...
+        return preferedMates;
+    }
+
+    public String toString() {
+        return getClass().getName() + "(" + name + ")";
+    }
+}
+
+class Woman extends Person {
+    private List<Man> suitors = new LinkedList<>();
```
Comment (Gustorn) [inline contents/stable_marriage_problem/code/java/stable-marriage.java]: Use `ArrayList`

File: contents/stable_marriage_problem/code/java/stable-marriage.java (hunk 2 above)
Comment (bsamseth) [reply contents/stable_marriage_problem/code/java/stable-marriage.java]: I very much agree, that is much cleaner!

File: contents/stable_marriage_problem/code/java/stable-marriage.java (hunk 3 above)
Comment (bsamseth) [reply contents/stable_marriage_problem/code/java/stable-marriage.java]: The reasoning was that removal from the front of the list would be efficient. But I agree, `ArrayList` is probably at least just as good.

**Comment by bsamseth** [comment]
@Gustorn Have you had time to look over the changes? Just realized this has been sitting untouched for some weeks now.

**Comment by Butt4cak3** [comment]
I reviewed the changes you made and found everything like @Gustorn requested it. I also went ahead and checked that everything displays properly on the website. I'll go ahead and merge this.

**Pull request merged**